zaza/openstack/charm_tests/ceph/rbd_mirror/tests.py

```diff
@@ -258,85 +347,170 @@ def create_volume_from_image(cinder, image, retry=20):
 class CephRBDMirrorControlledFailoverTest(CephRBDMirrorBase):
     """Encapsulate ``ceph-rbd-mirror`` controlled failover tests."""
 
+    def cinder_fail_over_fall_back(self):
+        """Validate controlled fail over and fall back via the Cinder API."""
+        session = openstack.get_overcloud_keystone_session()
+        cinder = openstack.get_cinder_session_client(session)
+        volume = self.create_cinder_volume(session, from_image=True)
+        self.wait_for_mirror_state(
+            'up+replaying',
+            check_entries_behind_master=True,
+            application_name=self.application_name + self.site_b_app_suffix,
+            model_name=self.site_b_model,
+            pools=['cinder-ceph'])
+        self.failover_cinder_volume_host(
+            cinder_client=cinder)
+        self.assertEqual(cinder.volumes.get(volume.id).status, 'available')
+        self.failover_cinder_volume_host(
+            cinder_client=cinder,
+            target_backend_id='default',
+            target_status='enabled',
+            target_replication_status='enabled')
+        self.assertEqual(cinder.volumes.get(volume.id).status, 'available')
+
     def test_fail_over_fall_back(self):
         """Validate controlled fail over and fall back."""
         site_a_pools, site_b_pools = self.get_pools()
+        site_a_action_params = {}
+        site_b_action_params = {}
+        if self.get_cinder_rbd_mirroring_mode() == 'image':
+            site_a_pools.remove('cinder-ceph')
```

Comment:
I'm getting lost here. It feels like this test requires expert knowledge to be understood. What I read is "if the mirroring mode is image, remove a pool called cinder-ceph". I'm lost and I don't understand what's going on here. I think we should add inline comments to make it possible for someone who isn't an expert here to vaguely understand what is going on :) Why is it necessary to remove pools in order to validate a failover? Are these pools a leftover from previous tests? Should they be cleaned up in these previous tests then?

Reply:
No.

I'm sorry for the opacity. The idea goes like this:

* When you have an environment with Ceph image-mirroring enabled, you want to do the fail-over / fail-back via Cinder API, **only** for the Cinder pool (because that's the reason image-mirroring makes sense in the first place).
* However, in the same environment, there might be other pools replicated (like Glance, for example). These pools need to be fail-overed / fail-back as usual, via the charm actions.

  This is exactly the main reason I added another optional parameter to the `ceph-rbd-mirror` actions, with the pools list. Because we want to exclude the `cinder-ceph` pool when running the Juju actions, since Cinder is enabled to do the `demote`, `promote` stuff.

That's the extra functionality here. The test is doing everything as before, but excluding the `cinder-ceph` pool, *only if* Cinder has image mirroring enabled (and we use the Cinder API in that case).

I'll make sure I add comments / docstring to clearly reflect this.